protocol/messenger_contact_requests_test.go

```diff
@@ -1276,3 +1276,398 @@ func (s *MessengerContactRequestSuite) TestPairedDevicesRemoveContact() {
+		},
+		"no messages",
+	)
+	s.Require().NoError(err)
+
+}
+
+// The scenario tested is as follow:
+// 1) Alice sends a contact request to Bob
+// 2) Bob accepts the contact request
+// 3) Bob goes offline
+// 4) Alice retracts the contact request
+// 5) Alice adds bob back to her contacts
+// 6) Bob goes online, they receive 4 and 5 in the wrong order
+func (s *MessengerContactRequestSuite) TestAliceOfflineRetractsAndAddsWrongOrder() {
+	// Alice sends a contact request to bob
+	alice1 := s.m
+
+	bob := s.newMessenger(s.shh)
+	_, err := bob.Start()
+	s.Require().NoError(err)
+
+	bobID := types.EncodeHex(crypto.FromECDSAPub(&bob.identity.PublicKey))
+	myID := types.EncodeHex(crypto.FromECDSAPub(&alice1.identity.PublicKey))
+
+	request := &requests.AddContact{
+		ID: types.Hex2Bytes(bobID),
+	}
+
+	_, err = alice1.AddContact(context.Background(), request)
+	s.Require().NoError(err)
+
+	// Wait for the message to reach its destination
+	resp, err := WaitOnMessengerResponse(
+		bob,
+		func(r *MessengerResponse) bool {
+			return len(r.Contacts) > 0 && len(r.Messages()) > 0 && len(r.ActivityCenterNotifications()) > 0
+		},
+		"no messages",
+	)
+
+	// Check contact request has been received
+	s.Require().NoError(err)
+
+	// Check the contact state is correctly set
+	s.Require().Len(resp.Contacts, 1)
+	s.Require().Equal(ContactRequestStateReceived, resp.Contacts[0].ContactRequestRemoteState)
+
+	// Bob accepts the contact request
+	_, err = bob.AcceptLatestContactRequestForContact(context.Background(), &requests.AcceptLatestContactRequestForContact{ID: types.Hex2Bytes(myID)})
+	s.Require().NoError(err)
+
+	// Alice receives the accepted confirmation
+	resp, err = WaitOnMessengerResponse(
+		alice1,
+		func(r *MessengerResponse) bool {
+			return len(r.Contacts) > 0
+		},
+		"no messages",
+	)
+	s.Require().NoError(err)
+
+	// Make sure we consider them a mutual contact, sender side
+	mutualContacts := alice1.MutualContacts()
+	s.Require().Len(mutualContacts, 1)
+
+	// Check the contact state is correctly set
+	s.Require().Len(resp.Contacts, 1)
+	s.Require().True(resp.Contacts[0].mutual())
+
+	_, err = alice1.RetractContactRequest(&requests.RetractContactRequest{ContactID: types.Hex2Bytes(bob.myHexIdentity())})
+	s.Require().NoError(err)
+
+	// adds bob again to her device
+	request = &requests.AddContact{
+		ID: types.Hex2Bytes(bobID),
+	}
+
+	_, err = alice1.AddContact(context.Background(), request)
+	s.Require().NoError(err)
+
+	// Get alice perspective of bob
+	bobFromAlice := alice1.AddedContacts()[0]
+
+	// Get bob perspective of alice
+	aliceFromBob := bob.MutualContacts()[0]
+
+	s.Require().NotNil(bobFromAlice)
+	s.Require().NotNil(aliceFromBob)
+
+	// We can't simulate out-of-order messages easily, so we need to do
+	// things manually here
+
+	result := aliceFromBob.ContactRequestPropagatedStateReceived(bobFromAlice.ContactRequestPropagatedState())
+	s.Require().True(result.newContactRequestReceived)
+
```

Comment:
I see a handful of these extra newlines at the end/beginning of function bodies. Have we ever looked at using https://github.com/mvdan/gofumpt? It's capable of auto formatting more stuff and it seems to be a drop-in replacement for gofmt. This tool can, among other *more useful* things, remove extraneous newlines at the end of function bodies, so all in all it could be a win.

Reply:
I'll look into it, thanks for the suggestion!